src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -71,6 +94,325 @@ private static Vector128<float> VectorSum(in Vector128<float> vector)
+            // likely be more efficient.
+            fixed (float* pSrcStart = &src.Items[0])
+            fixed (float* pDstStart = &dst.Items[0])
+            fixed (float* pMatStart = &mat.Items[0])
+            fixed (int* pposSrc = &rgposSrc[0])
+            {
+                float* psrc = Ptr(src, pSrcStart);
+                float* pdst = Ptr(dst, pDstStart);
+                float* pmat = Ptr(mat, pMatStart);
+
+                int* pposMin = pposSrc + iposMin;
+                int* pposEnd = pposSrc + iposEnd;
+                float* pDstEnd = pdst + crow;
+                float* pm0 = pmat - posMin;
+                float* pSrcCurrent = psrc - posMin;
+                float* pDstCurrent = pdst;
+
+                while (pDstCurrent < pDstEnd)
+                {
+                    float* pm1 = pm0 + ccol;
+                    float* pm2 = pm1 + ccol;
+                    float* pm3 = pm2 + ccol;
+                    Vector128<float> res = Sse.SetZeroVector128();
+
+                    int* ppos = pposMin;
+
+                    while (ppos < pposEnd)
+                    {
+                        int col = *ppos;
+                        Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
+                        Vector128<float> x2 = Sse.SetAllVector128(pSrcCurrent[col]);
+                        x2 = Sse.Multiply(x2, x1);
+                        res = Sse.Add(res, x2);
+
+                        ppos++;
+                    }
+
+                    if (add)
+                    {
+                        res = Sse.Add(res, Sse.LoadAlignedVector128(pDstCurrent));
+                    }
+                    Sse.StoreAligned(pDstCurrent, res);
+
+                    pDstCurrent += 4;
+                    pm0 += 4 * ccol;
+                }
+            }
+        }
+
+        internal static unsafe void MatMulTranA(bool add, AlignedArray mat, AlignedArray src, AlignedArray dst, int crow, int ccol)
+        {
+            Contracts.Assert(Compat(mat));
+            Contracts.Assert(Compat(src));
+            Contracts.Assert(Compat(dst));
+
+            fixed (float* pSrcStart = &src.Items[0])
+            fixed (float* pDstStart = &dst.Items[0])
+            fixed (float* pMatStart = &mat.Items[0])
+            {
+                float* psrc = Ptr(src, pSrcStart);
+                float* pdst = Ptr(dst, pDstStart);
+                float* pmat = Ptr(mat, pMatStart);
+
+                float* pSrcEnd = psrc + ccol;
+                float* pDstEnd = pdst + crow;
+                float* pSrcCurrent = psrc;
+                float* pDstCurrent = pdst;
+                float* pMatCurrent = pmat;
+
+                if (!add)
+                {
+                    Vector128<float> x01 = Sse.LoadAlignedVector128(pSrcCurrent);
+                    // Replicate each slot of x01 into its own register.
+                    Vector128<float> x11 = Sse.Shuffle(x01, x01, 0x55);
```

Comment:
A comment about what each of these is extracting would be useful:
* `0x00` == `X`
* `0x55` == `Y`
* `0xAA` == `Z`
* `0xFF` == `W`

Reply:
Thanks and adopted the change.  May I know if there is any preference/convention for using `XYZW` instead of `ABCD`?  I used `ABCD` in comments of private functions earlier thinking that all users can always infer the correct elements from their alphabetical order, but not sure if it's the best way.